vortex-cxx/cpp/include/vortex/file.hpp

```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright the Vortex contributors
+
+#pragma once
+
+#include <nanoarrow/common/inline_types.h>
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <string_view>
+#include <utility>
+
+namespace vortex {
+
+class ScanBuilder {
+public:
+    ScanBuilder(ScanBuilder &&other) noexcept;
+    ScanBuilder &operator=(ScanBuilder &&other) noexcept;
+    ~ScanBuilder();
+
+    ScanBuilder(const ScanBuilder &) = delete;
+    ScanBuilder &operator=(const ScanBuilder &) = delete;
+
+    /// Set the filter on the scan builder.
+    ScanBuilder &SetFilter(std::string_view filter);
+
+    /// Set the limit on the number of rows to scan.
+    ScanBuilder &SetLimit(uint64_t limit);
+
+    /// Consume the scan builder to a stream of record batches.
+    /// The scan builder is consumed and cannot be used after this call.
+    ArrowArrayStream IntoStream();
+
+    /// Consume the scan builder to an Arrow array and schema.
+    /// The scan builder is consumed and cannot be used after this call.
+    std::pair<ArrowArray, ArrowSchema> IntoArray();
+
+private:
+    friend class VortexFile;
+
+    struct Impl;
+    explicit ScanBuilder(std::unique_ptr<Impl> impl);
+
+    std::unique_ptr<Impl> impl_;
+};
```

Comment:
can this live in a different file `scan`?

Reply:
Done. I also reorg the directory so that `include/` is public headers and `src/` is for private headers and source files.

Reply:
It is in. Commit here: https://github.com/vortex-data/vortex/pull/3792/commits/9482b526f27f2190cbdeb64861a6fbc8c0f614b7